File: DockerHub/coap_data_sender/coap_data_sender.py
import asyncio
import csv
import json
import os
import sys
from aiocoap import *

COAP_SERVER_URL = os.getenv('COAP_SERVER_URL')
SEND_INTERVAL = int(os.getenv('SEND_INTERVAL', 1))
CSV_URL = os.getenv('CSV_URL')
INSTANCE_DATA = os.getenv('INSTANCE_DATA')

print("Variáveis de Ambiente Recebidas:")
print(f"COAP_SERVER_URL: {COAP_SERVER_URL}")
print(f"SEND_INTERVAL: {SEND_INTERVAL}")
print(f"CSV_URL: {CSV_URL}")
print(f"INSTANCE_DATA: {INSTANCE_DATA}")

def download_csv(url):
    """Baixa o CSV da URL fornecida e o processa em uma lista de dicionários."""
    try:
        import requests
        response = requests.get(url)
        response.raise_for_status()
        lines = response.text.splitlines()
        reader = csv.DictReader(lines)
        return list(reader)
    except Exception as e:
        print(f"Erro ao baixar ou processar o CSV: {e}")
        sys.exit(1)

async def send_data_coap(data, coap_url):
    """Envia os dados para o servidor CoAP."""
    context = await Context.create_client_context()
    payload = json.dumps(data).encode('utf-8')
    request = Message(code=POST, payload=payload, uri=coap_url)

    try:
        print(f"Tentando enviar dados CoAP: {data}")
        response = await context.request(request).response
        print(f"Dados enviados com sucesso! Resposta: {response.code}, {response.payload.decode()}")
    except Exception as e:
        print(f"Falha ao enviar dados CoAP: {e}")

async def start_sending_data_coap(data_list, coap_url, unique_id, street):
    """Envia os dados do CSV periodicamente via CoAP."""
    for data in data_list:
        data_to_send = {
            "type": "consumption",
            "id": unique_id,
            "street": street,
            **data
        }

        print("Enviando dados via CoAP:")
        print(data_to_send)

        await send_data_coap(data_to_send, coap_url)
        await asyncio.sleep(SEND_INTERVAL)

if __name__ == '__main__':
    if not COAP_SERVER_URL or not CSV_URL or not INSTANCE_DATA:
        print("Erro: As variáveis de ambiente 'CSV_URL', 'COAP_SERVER_URL', e 'INSTANCE_DATA' devem ser definidas.")
        sys.exit(1)

    # Carrega os dados da instância
    instance_data = json.loads(INSTANCE_DATA)
    unique_id = instance_data.get("id")
    street = instance_data.get("street")

    print("Dados da Instância Carregados:")
    print(f"ID Único: {unique_id}")
    print(f"Endereço: {street}")

    # Baixa e processa o CSV
    print("Baixando e processando o CSV...")
    data_list = download_csv(CSV_URL)

    print("Iniciando processo de envio CoAP...")
    asyncio.run(start_sending_data_coap(data_list, COAP_SERVER_URL, unique_id, street))
    print("Medidor CoAP encerrado com sucesso.")

File: DockerHub/coap_data_sender/coap_data_sender.js
const coap = require('coap');
const axios = require('axios');

// Carregar variáveis de ambiente
const COAP_SERVER_URL = process.env.COAP_SERVER_URL;
const CSV_URL = process.env.CSV_URL;
const INSTANCE_DATA = process.env.INSTANCE_DATA;
const SEND_INTERVAL = parseInt(process.env.SEND_INTERVAL || "1", 10);

if (!COAP_SERVER_URL || !CSV_URL || !INSTANCE_DATA) {
    console.error("Erro: Variáveis de ambiente obrigatórias não definidas.");
    process.exit(1);
}

let instanceData;
try {
    instanceData = JSON.parse(INSTANCE_DATA);
} catch (err) {
    console.error("Erro ao processar INSTANCE_DATA:", err);
    process.exit(1);
}

// Ajustar a URL para garantir que não tenha "/receive_data" indevidamente
const coapUrl = new URL(COAP_SERVER_URL.replace("/receive_data", ""));
const uniqueId = instanceData.id;
const street = instanceData.street;

console.log("### Configurações ###");
console.log(`COAP_SERVER_URL: ${coapUrl.href}`);
console.log(`Hostname: ${coapUrl.hostname}`);
console.log(`Porta: ${coapUrl.port}`);
console.log(`CSV_URL: ${CSV_URL}`);
console.log(`INSTANCE_DATA: ${JSON.stringify(instanceData)}`);
console.log(`SEND_INTERVAL: ${SEND_INTERVAL}s`);

// Função para registrar o medidor no nó primário via CoAP
async function registerMeter(meterType) {
    return new Promise((resolve) => {
        const req = coap.request({
            method: "POST",
            hostname: coapUrl.hostname,
            pathname: "/register_meter",
            port: coapUrl.port,
        });

        const payload = JSON.stringify({ id: uniqueId, street: street, type: meterType });
        req.setOption("Content-Format", "application/json");
        req.write(payload);

        console.log(`Registrando medidor na URL CoAP: ${coapUrl.href}/register_meter com payload:`, payload);

        req.on("response", (res) => {
            const responseText = res.payload.toString();
            console.log(`Resposta do nó primário: ${res.code} - ${responseText}`);
            if (res.code.toString().startsWith("2.0")) {
                console.log(`Medidor ${uniqueId} (${meterType}) registrado com sucesso.`);
                resolve(true);
            } else {
                console.warn(`Falha no registro do medidor ${uniqueId}. Tentando novamente...`);
                resolve(false);
            }
        });

        req.on("error", (err) => {
            console.error("Erro ao registrar medidor:", err);
            resolve(false);
        });

        req.end();
    });
}

// Função para baixar e processar o CSV
async function downloadCsv(url) {
    try {
        console.log(`Baixando CSV de ${url}...`);
        const response = await axios.get(url);
        const lines = response.data.split("\n");

        console.log("Primeiras 5 linhas do CSV baixado:");
        for (let i = 0; i < Math.min(5, lines.length); i++) {
            console.log(lines[i]);
        }

        // Processar CSV
        const headers = lines.shift().split(";").map(header => header.trim());
        const data = lines
            .filter(line => line.trim())
            .map(line => {
                const values = line.split(";").map(value => value.trim());
                return headers.reduce((acc, header, index) => {
                    acc[header] = values[index] || null;
                    return acc;
                }, {});
            });

        // Detectar tipo do medidor
        let meterType = "desconhecido";
        if (headers.includes("consumption_m3_per_hour")) {
            meterType = "water";
        } else if (headers.includes("consumption_kwh_per_hour")) {
            meterType = "energy";
        }

        console.log(`Tipo de medidor detectado: ${meterType}`);
        return { data, meterType };
    } catch (error) {
        console.error("Erro ao baixar ou processar o CSV:", error);
        process.exit(1);
    }
}

// Função para enviar dados via CoAP
function sendDataCoap(data) {
    const req = coap.request({
        method: "POST",
        hostname: coapUrl.hostname,
        pathname: "/receive_data",  // 🔧 CORRIGIDO!
        port: coapUrl.port, 
    });

    req.setOption("Content-Format", "application/json");
    req.write(JSON.stringify(data));

    req.on("response", (res) => {
        console.log(`Resposta recebida: ${res.code} - ${res.payload.toString()}`);
    });

    req.on("error", (err) => {
        console.error("Erro ao enviar dados via CoAP:", err);
    });

    req.end();
}

// Função principal
async function main() {
    console.log("Baixando e processando o CSV...");
    const { data, meterType } = await downloadCsv(CSV_URL);

    console.log("Registrando medidor no nó primário...");
    while (!(await registerMeter(meterType))) {
        console.log("Tentando novamente em 5 segundos...");
        await new Promise(resolve => setTimeout(resolve, 5000));
    }

    for (const dataEntry of data) {
        if (!dataEntry.date || !dataEntry.time) {
            console.warn("Aviso: Linha com dados incompletos ignorada:", dataEntry);
            continue;
        }

        const dataToSend = {
            type: meterType,
            id: uniqueId,
            street: street,
            ...dataEntry,
        };

        console.log("Enviando dados:", dataToSend);
        sendDataCoap(dataToSend);

        await new Promise(resolve => setTimeout(resolve, SEND_INTERVAL * 1000));
    }

    console.log("Envio concluído com sucesso!");
}

main().catch((err) => {
    console.error("Erro fatal:", err);
    process.exit(1);
});
